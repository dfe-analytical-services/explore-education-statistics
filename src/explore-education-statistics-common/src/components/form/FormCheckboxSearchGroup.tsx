import { FormFieldsetProps } from '@common/components/form/FormFieldset';
import FormTextSearchInput from '@common/components/form/FormTextSearchInput';
import { FormFieldset } from '@common/components/form/index';
import useMounted from '@common/hooks/useMounted';
import formatPretty from '@common/utils/number/formatPretty';
import React, { useState } from 'react';
import FormCheckboxGroup, {
  BaseFormCheckboxGroup,
  FormCheckboxGroupProps,
} from './FormCheckboxGroup';
import styles from './FormCheckboxSearchGroup.module.scss';
import VisuallyHidden from '../VisuallyHidden';

export interface FormCheckboxSearchGroupProps extends FormCheckboxGroupProps {
  maxSearchResults?: number;
  searchHelpText?: string;
  searchLabel?: string;
  searchOnly?: boolean;
}

const FormCheckboxSearchGroup = ({
  searchLabel,
  ...props
}: FormCheckboxSearchGroupProps) => {
  const { isMounted } = useMounted();

  const [searchTerm, setSearchTerm] = useState('');

  if (!isMounted) {
    return <FormCheckboxGroup {...props} />;
  }

  const {
    error,
    hint,
    groupLabel,
    id,
    legend,
    legendHidden,
    legendSize,
    maxSearchResults = 500,
    name,
    options = [],
    searchHelpText,
    searchOnly = false,
    value = [],
    onFieldsetFocus,
    onFieldsetBlur,
    ...groupProps
  } = props;

  const minSearchCharacters = searchOnly ? 3 : 0;

  const fieldsetProps: FormFieldsetProps = {
    id,
    legend,
    legendHidden,
    legendSize,
    hint,
    error,
    onFocus: onFieldsetFocus,
    onBlur: onFieldsetBlur,
  };

  let filteredOptions = searchOnly
    ? options.filter(option => value.includes(option.value))
    : options;

  if (searchTerm) {
    // Search for URN in the hint field if search term is a number.
    if (searchOnly && !Number.isNaN(Number(searchTerm))) {
      filteredOptions = options.filter(
        option =>
          option.hint?.toString().includes(searchTerm) ||
          value.includes(option.value),
      );
    } else {
      filteredOptions = options.filter(
        option =>
          option.label
            .toLowerCase()
            .includes(searchTerm.trim().toLowerCase()) ||
          value.includes(option.value),
      );
    }
  }

  const getResultsMessage = () => {
    const numResults = filteredOptions.length;
    if (!searchOnly && numResults === 0) {
      return <p>No options available.</p>;
    }
    if (searchOnly) {
      if (numResults === 0) {
        return (
          <p>
            {searchHelpText || 'Search above and select at least one option.'}
          </p>
        );
      }
      if (numResults > maxSearchResults) {
        return (
          <p>
            {formatPretty(numResults)} results found. Please refine your search
            to view options.
          </p>
        );
      }
    }
    return null;
  };

  const showResults = !searchOnly || filteredOptions.length <= maxSearchResults;

  return (
    <FormFieldset {...fieldsetProps} useFormId={false}>
      <FormTextSearchInput
        id={`${id}-search`}
        name={`${name}-search`}
        label={
          searchLabel || (
            <>
              Search options
              {groupLabel && (
                <VisuallyHidden>{` for ${groupLabel}`}</VisuallyHidden>
              )}
            </>
          )
        }
        width={20}
        onChange={event => {
          if (
            event.target.value.length >= minSearchCharacters ||
            !event.target.value.length
          ) {
            setSearchTerm(event.target.value);
          }
        }}
        onKeyPress={event => {
          if (event.key === 'Enter') {
            event.preventDefault();
          }
        }}
      />

      <div className={styles.optionsContainer}>
        {filteredOptions.length > 0 && (
          <span
            aria-live="polite"
            aria-atomic
            className="govuk-visually-hidden"
          >
            {`${filteredOptions.length} option${
              filteredOptions.length > 1 ? 's' : ''
            } found`}
          </span>
        )}
        {showResults && filteredOptions.length > 0 && (
          <BaseFormCheckboxGroup
            {...groupProps}
            id={`${id}-options`}
            value={value}
            name={name}
            groupLabel={groupLabel}
            options={filteredOptions}
            small
          />
        )}
        {getResultsMessage()}
      </div>
    </FormFieldset>
  );
};

export default FormCheckboxSearchGroup;
