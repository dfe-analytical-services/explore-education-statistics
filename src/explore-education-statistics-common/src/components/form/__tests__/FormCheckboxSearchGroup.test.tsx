import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import FormCheckboxSearchGroup from '../FormCheckboxSearchGroup';

jest.mock('lodash/debounce');

describe('FormCheckboxSearchGroup', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  describe('without searchOnly', () => {
    test('renders list of checkboxes in correct order', () => {
      const { container } = render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          value={[]}
          options={[
            { label: 'Test checkbox 1', value: '1' },
            { label: 'Test checkbox 2', value: '2' },
            { label: 'Test checkbox 3', value: '3' },
          ]}
        />,
      );

      const checkboxes = screen.getAllByLabelText(
        /Test checkbox/,
      ) as HTMLInputElement[];

      expect(checkboxes).toHaveLength(3);
      expect(checkboxes[0]).toHaveAttribute('value', '1');
      expect(checkboxes[1]).toHaveAttribute('value', '2');
      expect(checkboxes[2]).toHaveAttribute('value', '3');

      expect(container.innerHTML).toMatchSnapshot();
    });

    test('providing a search term renders only relevant checkboxes', async () => {
      jest.useFakeTimers();

      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          value={[]}
          options={[
            { label: 'Test checkbox 1', value: '1' },
            { label: 'Test checkbox 2', value: '2' },
            { label: 'Test checkbox 3', value: '3' },
          ]}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');

      userEvent.type(searchInput, '2');

      jest.runAllTimers();

      const checkboxes = screen.getAllByLabelText(
        /Test checkbox/,
      ) as HTMLInputElement[];

      expect(checkboxes).toHaveLength(1);
      expect(checkboxes[0]).toHaveAttribute('value', '2');
    });

    test('providing a search term does not remove checkboxes that have already been checked', async () => {
      jest.useFakeTimers();

      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          value={['1']}
          options={[
            { label: 'Test checkbox 1', value: '1' },
            { label: 'Test checkbox 2', value: '2' },
            { label: 'Test checkbox 3', value: '3' },
          ]}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');

      userEvent.type(searchInput, '2');

      jest.runAllTimers();

      const checkboxes = screen.getAllByLabelText(
        /Test checkbox/,
      ) as HTMLInputElement[];

      expect(checkboxes).toHaveLength(2);
      expect(checkboxes[0]).toHaveAttribute('value', '1');
      expect(checkboxes[1]).toHaveAttribute('value', '2');
    });

    test('renders the checkboxes when the number of options is above the maxSearchResults', () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          maxSearchResults={2}
          value={[]}
          options={[
            { label: 'Test checkbox 1', value: '1' },
            { label: 'Test checkbox 2', value: '2' },
            { label: 'Test checkbox 3', value: '3' },
          ]}
        />,
      );

      const checkboxes = screen.getAllByLabelText(
        /Test checkbox/,
      ) as HTMLInputElement[];

      expect(checkboxes).toHaveLength(3);
      expect(checkboxes[0]).toHaveAttribute('value', '1');
      expect(checkboxes[1]).toHaveAttribute('value', '2');
      expect(checkboxes[2]).toHaveAttribute('value', '3');
    });

    test('renders the no options message when there are no options', () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          value={[]}
          options={[]}
        />,
      );

      expect(screen.getByText('No options available.')).toBeInTheDocument();
    });
  });

  describe('with searchOnly', () => {
    const testSearchOnlyOptions = [
      {
        label: 'Test checkbox 1',
        value: '1',
        hint: 'URN: 000001; Local authority: LA 1',
      },
      {
        label: 'Test checkbox 2',
        value: '2',
        hint: 'URN: 000002; Local authority: LA 1',
      },
      {
        label: 'Test checkbox 3',
        value: '3',
        hint: 'URN: 000003; Local authority: LA 1',
      },
      {
        label: 'Test 4',
        value: '4',
        hint: 'URN: 000004; Local authority: LA 1',
      },
    ];
    test('does not render the checkboxes by default', () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={testSearchOnlyOptions}
        />,
      );

      expect(screen.queryAllByRole('checkbox')).toHaveLength(0);
    });

    test('renders checkboxes for selected options', () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          searchOnly
          value={['2']}
          options={testSearchOnlyOptions}
        />,
      );

      const checkboxes = screen.getAllByRole('checkbox');
      expect(checkboxes).toHaveLength(1);
      expect(checkboxes[0]).toHaveAttribute('value', '2');
      expect(checkboxes[0]).toEqual(screen.getByLabelText('Test checkbox 2'));
      expect(checkboxes[0]).toBeChecked();
    });

    test('renders checkboxes that match the search term', async () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={testSearchOnlyOptions}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');
      userEvent.type(searchInput, 'checkbox');

      await waitFor(() => {
        expect(screen.getByText('3 options found')).toBeInTheDocument();
      });

      const checkboxes = screen.getAllByRole('checkbox');
      expect(checkboxes).toHaveLength(3);
      expect(checkboxes[0]).toHaveAttribute('value', '1');
      expect(checkboxes[0]).toEqual(screen.getByLabelText('Test checkbox 1'));
      expect(checkboxes[0]).not.toBeChecked();

      expect(checkboxes[1]).toHaveAttribute('value', '2');
      expect(checkboxes[1]).toEqual(screen.getByLabelText('Test checkbox 2'));
      expect(checkboxes[1]).not.toBeChecked();

      expect(checkboxes[2]).toHaveAttribute('value', '3');
      expect(checkboxes[2]).toEqual(screen.getByLabelText('Test checkbox 3'));
      expect(checkboxes[1]).not.toBeChecked();
    });

    test('renders checkboxes where the URN in the hint matches the search term', async () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={testSearchOnlyOptions}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');
      userEvent.type(searchInput, '000002');

      await waitFor(() => {
        expect(screen.getByText('1 option found')).toBeInTheDocument();
      });

      const checkboxes = screen.getAllByRole('checkbox');

      expect(checkboxes[0]).toHaveAttribute('value', '2');
      expect(checkboxes[0]).toEqual(screen.getByLabelText('Test checkbox 2'));
      expect(checkboxes[0]).not.toBeChecked();
    });

    test('does not search when the search term has fewer than 3 characters', async () => {
      jest.useFakeTimers();
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={testSearchOnlyOptions}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');
      userEvent.type(searchInput, 'ch');

      jest.runAllTimers();

      expect(screen.queryAllByRole('checkbox')).toHaveLength(0);
    });

    test('renders the too many results message and no checkboxes when there are more than the maximum results', async () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          maxSearchResults={2}
          searchHelpText="This is the help text"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={testSearchOnlyOptions}
        />,
      );

      const searchInput = screen.getByLabelText('Search options');
      userEvent.type(searchInput, 'checkbox');

      await waitFor(() => {
        expect(
          screen.getByText(
            '3 results found. Please refine your search to view options.',
          ),
        ).toBeInTheDocument();
      });

      expect(screen.queryAllByRole('checkbox')).toHaveLength(0);
    });

    test('renders the search help when text there are no options', () => {
      render(
        <FormCheckboxSearchGroup
          name="testCheckboxes"
          id="test-checkboxes"
          legend="Choose options"
          maxSearchResults={2}
          searchHelpText="This is the help text"
          searchLabel="Search options"
          searchOnly
          value={[]}
          options={[]}
        />,
      );

      expect(screen.getByText('This is the help text')).toBeInTheDocument();
    });
  });

  test('does not throw error if search term that is invalid regex is used', async () => {
    jest.useFakeTimers();

    render(
      <FormCheckboxSearchGroup
        name="testCheckboxes"
        id="test-checkboxes"
        legend="Choose options"
        searchLabel="Search options"
        value={[]}
        options={[
          { label: 'Test checkbox 1', value: '1' },
          { label: 'Test checkbox 2', value: '2' },
          { label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    const searchInput = screen.getByLabelText('Search options');

    userEvent.type(searchInput, '[');

    expect(() => {
      jest.runAllTimers();
    }).not.toThrow();

    const checkboxes = screen.queryAllByLabelText(
      /Test checkbox/,
    ) as HTMLInputElement[];

    expect(checkboxes).toHaveLength(0);
  });
});
