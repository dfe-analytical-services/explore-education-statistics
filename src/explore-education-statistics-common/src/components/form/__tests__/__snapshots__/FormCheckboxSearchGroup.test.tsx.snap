// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxSearchGroup without searchOnly renders list of checkboxes in correct order 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      Choose options
    </legend>
    <label class="govuk-label"
           for="test-checkboxes-search"
    >
      Search options
    </label>
    <input name="testCheckboxes-search"
           class="govuk-input searchInput govuk-input--width-20"
           id="test-checkboxes-search"
           type="text"
           value
    >
    <div class="optionsContainer">
      <span aria-live="polite"
            aria-atomic="true"
            class="govuk-visually-hidden"
      >
        3 options found
      </span>
      <div class="govuk-checkboxes govuk-checkboxes--small">
        <div class="govuk-checkboxes__item"
             data-testid="Checkbox item for Test checkbox 1"
        >
          <input class="govuk-checkboxes__input"
                 id="test-checkboxes-options-1"
                 name="testCheckboxes"
                 type="checkbox"
                 value="1"
          >
          <label class="govuk-label govuk-checkboxes__label"
                 for="test-checkboxes-options-1"
          >
            Test checkbox 1
          </label>
        </div>
        <div class="govuk-checkboxes__item"
             data-testid="Checkbox item for Test checkbox 2"
        >
          <input class="govuk-checkboxes__input"
                 id="test-checkboxes-options-2"
                 name="testCheckboxes"
                 type="checkbox"
                 value="2"
          >
          <label class="govuk-label govuk-checkboxes__label"
                 for="test-checkboxes-options-2"
          >
            Test checkbox 2
          </label>
        </div>
        <div class="govuk-checkboxes__item"
             data-testid="Checkbox item for Test checkbox 3"
        >
          <input class="govuk-checkboxes__input"
                 id="test-checkboxes-options-3"
                 name="testCheckboxes"
                 type="checkbox"
                 value="3"
          >
          <label class="govuk-label govuk-checkboxes__label"
                 for="test-checkboxes-options-3"
          >
            Test checkbox 3
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;
